### DISCUSSION TITLE
react-native-reanimated new architecture adoption

### DISCUSSION THREAD

**Discussion by tomekzaw** [reactwg/react-native-new-architecture#14] [Libraries]
We are happy to announce that Reanimated supports Fabric as of 3.0.0-rc.0 release. 🎉 

For more details, check out our [Announcing Reanimated 3](https://blog.swmansion.com/announcing-reanimated-3-16167428c5f7) blog post.

----

React Native Reanimated integrates quite closely with the renderer system so we needed to rewrite some parts specifically for Fabric and that's exactly what we have been doing for the past few months.

Here's the current state of feature migration:
* ✅ secondary JS VM & worklets
* ✅ animating UI props (e.g. `opacity` or `backgroundColor`) 
* ✅ animating native props (e.g. `width` or `margin`)
* ✅ `useAnimatedScrollHandler`
* ✅ `measure` & `scrollTo` in worklets
* Layout Animations – TODO
* ✅ integration with Gesture Handler
* ✅ simultaneous support for Paper and Fabric, autoinstall etc. – should be easy since we've already done it for `react-native-screens` and `react-native-gesture-handler` 
* profiling & optimization – after RC

Pull requests to React Native:
* https://github.com/facebook/react-native/pull/33500 - we've had some troubles with `dynamic_cast<ShadowNodeWrapper>` due to RTTI not working properly between shared library boundaries and this PR to React Native core enables third-party libraries to integrate with C++ part of Fabric (edit: merged, thanks!)
* https://github.com/facebook/react-native/pull/33510 - this PR should eliminate some major difficulties blocking us from implementing Layout Animations for Fabric (edit: merged, thanks!)
* https://github.com/facebook/react-native/pull/33688 - this PR allows us to easily obtain `UIManager` and also intercept and elegantly handle events in C++ (edit: merged, thanks!)

With Fabric release we also dropped support for Reanimated 1.

**Comment by cortinico** [comment]
> Right now we're stuck on Android build config but once we resolve this issue we'll prepare the Fabric branch so other early adopters of the new architecture can check it out.

Is this something you can share more @tomekzaw or that we can support you in some form?

**Reply by tomekzaw** [reply]
@cortinico Sure! In Reanimated we need to use C++ ShadowNodes to calculate layout while animating props like `width` or `margin`.

We use `findHostInstance_DEPRECATED(ref)._internalInstanceHandle.stateNode.node` to obtain `ShadowNodeWrapper` (`jsi::HostObject`) from React ref in JavaScript. Once we have a ShadowNodeWrapper, we pass it to our JSI function.

In C++, we call [`shadowNodeFromValue`](https://github.com/facebook/react-native/blob/8bd3edec88148d0ab1f225d2119435681fbbba33/ReactCommon/react/renderer/uimanager/primitives.h#L43-L53) to convert `jsi::Value` to `ShadowNode::Shared`. This works fine on iOS but unfortunately crashes on Android (more specifically, on [this](https://github.com/facebook/react-native/blob/8bd3edec88148d0ab1f225d2119435681fbbba33/ReactCommon/jsi/jsi/jsi-inl.h#L180) line). The assert fails because `std::dynamic_pointer_cast` inside `isHostObject<T>` implementation returns `nullptr`.

It took us a while to investigate what is the possible root cause. We have a `ShadowNodeWrapper` instance that has been created in `libreact_render_uimanager.so` but we use `dynamic_cast<ShadowNodeWrapper>` in Reanimated codebase which gets compiled into `libreanimated.so` and most likely RTTI doesn't seem to work outside shared library boundaries (although `-frtti` is set both libraries).

> The module that constructed an object and the one performing the dynamic_cast might hold their own, otherwise identical, RTTI symbols for a type. This makes them appear as different types when the `dynamic_cast` is performed in a different module, causing `dynamic_cast` to unexpectedly return `nullptr`, or throw `std::bad_cast` for references. https://stackoverflow.com/a/56835475

Also, it doesn't matter if we use `shadowNodeFromValue` or `getHostObject<ShadowNodeWrapper>` directly, because both functions are `inline` (the first is defined in `primitives.h` and the latter in `jsi-inl.h`), so effectively they are embedded in the machine code in each shared library that `#include`s them instead of being compiled once and linked.

If `shadowNodeFromValue` wasn't `static inline`, third-party libraries could link to its implementation in `libreact_render_uimanager.so` (where `dynamic_cast` works fine). But it makes sense to leave this function `inline` due to performance reasons as it's frequently called from UIManagerBinding.

As an experiment, I've declared and implemented the following function in some `.h` and `.cpp` files in `react/renderer/uimanager` directory in React Native sources:
```c++
ShadowNode::Shared shadowNodeFromValueForThirdPartyLibs(
    jsi::Runtime &runtime,
    jsi::Value const &value) { // exported symbol
  return shadowNodeFromValue(runtime, value); // inlined
}
```
This way Reanimated can link to it since it's an exported symbol while the compiler can still inline `shadowNodeFromValue` in React Native internals as it does now. I know it's a workaround and most likely something is missing in our CMake config but this function might come handy for other third-party library developers as well.

By the way, is there any convenient way to obtain C++ `UIManager` and `ShadowTreeRegistry`? Currently we use `UIManagerBinding::getBinding(rt)` and some wild `reinterpret_cast`s to access private fields of UIManager but looking forward to making it more elegant.

**Reply by tomekzaw** [reply]
Just for the record, here's a PR to React Native that resolves this issue for all third-party libraries:
* https://github.com/facebook/react-native/pull/33500

**Reply by ShikaSD** [reply]
> By the way, is there any convenient way to obtain C++ UIManager and ShadowTreeRegistry? Currently we use UIManagerBinding::getBinding(rt) and some wild reinterpret_casts to access private fields of UIManager but looking forward to making it more elegant.

What are you trying to achieve with those? These APIs are designed to be private, but maybe we can expose required functionality somehow?

**Reply by ShikaSD** [reply]
> > We use findHostInstance_DEPRECATED(ref)._internalInstanceHandle.stateNode.node to obtain ShadowNodeWrapper (jsi::HostObject) from React ref in JavaScript. Once we have a ShadowNodeWrapper, we pass it to our JSI function.

Do you animate those props from JS or somehow natively? `_internalInstanceHandle` is an implementation detail and it is likely to break at some point. Code pointer might help here as well!

**Reply by tomekzaw** [reply]
@ShikaSD Here's more context: Reanimated v2 creates a secondary JS VM that is used only from the UI thread to synchronously run user-defined JS functions (so-called worklets). These functions are meant to calculate new values of animated props and return them as JS object (e.g. `{ opacity: 0.5, width: 100 }`) on each animation frame. Finally, these "animated styles" are applied via native part of RN internals (but without accessing main JS runtime). This allows library users to implement animations in pure JavaScript and still have them run smoothly with native performance, especially when the main JS VM is busy. 

Some properties (like `opacity` or `backgroundColor`) don't affect the size of the component but some do (e.g. `width` or `margin`). We call these "non-layout" and "layout" properties, respectively, because if any property from the latter group gets updated, we need to re-calculate the layout. (Historically, these groups were called "UI props" and "native props" but such naming is confusing.) We scan the prop names [here](https://github.com/software-mansion/react-native-reanimated/blob/3106c8e0270c336ed39a0523ea10564e23e860b2/Common/cpp/NativeModules/NativeReanimatedModule.cpp#L377-L407).

On Fabric, we use `synchronouslyUpdateViewOnUIThread` from RN internals to update non-layout props directly onto platform-specific native views (see [here](https://github.com/software-mansion/react-native-reanimated/blob/3106c8e0270c336ed39a0523ea10564e23e860b2/android/src/main/java/com/swmansion/reanimated/NodesManager.java#L531-L534) for Android and [here](https://github.com/software-mansion/react-native-reanimated/blob/3106c8e0270c336ed39a0523ea10564e23e860b2/ios/REANodesManager.mm#L315-L322) for iOS) and it works really well.

For updating layout props, the plan is to implement a `ShadowTreeCommitTransaction` that will also calculate the layout and mount the changes. Currently, we obtain `ShadowNodeWrapper` (as `jsi::HostObject`) from React ref ([here](https://github.com/software-mansion/react-native-reanimated/blob/3106c8e0270c336ed39a0523ea10564e23e860b2/src/reanimated2/getShadowNodeFromRef.ts#L18-L22)) on the main JS runtime. (We have some logic that makes it possible to share this `std::shared_ptr<ShadowNodeWrapper>` with the secondary JS VM as a different JS object but with the same underlying `std::shared_ptr<ShadowNodeWrapper>` because it is runtime-independent.) Now, on each animation frame, on the UI thread we batch all updates in `std::vector` (see [here](https://github.com/software-mansion/react-native-reanimated/blob/3106c8e0270c336ed39a0523ea10564e23e860b2/Common/cpp/NativeModules/NativeReanimatedModule.cpp#L405)) and at the end we call `performOperations` method which commits the transaction [here](https://github.com/software-mansion/react-native-reanimated/blob/3106c8e0270c336ed39a0523ea10564e23e860b2/Common/cpp/NativeModules/NativeReanimatedModule.cpp#L429-L461) (we'll definitely need to improve this implementation later on as it clones the root node *n*+1 times where *n* is the number of updates).

> What are you trying to achieve with those? These APIs are designed to be private, but maybe we can expose required functionality somehow?

We need `UIManager` to access `ShadowTreeRegistry` and create `PropsParserContext` in order to update layout props.

> Do you animate those props from JS or somehow natively?

Natively, from UI thread (layout must also be calculated on the UI thread, like pictured [here](https://reactnative.dev/architecture/threading-model#render-in-the-ui-thread)).

> `_internalInstanceHandle` is an implementation detail and it is likely to break at some point.

That's true, and using `findHostInstance_DEPRECATED` doesn't sound like a great idea as well. It would be great if there was an official way to get `ShadowNodeWrapper` from React ref. I'm pretty sure that other libraries will need this functionality as well. For example, in React Native Gesture Handler we also need to access ShadowNode in order to see if the corresponding view will get flattened by checking `ShadowNodeTraits::FormsStackingContext` (see [here](https://github.com/software-mansion/react-native-gesture-handler/blob/242c9af654428be7577067c8fd3c760e202008ac/src/handlers/gestures/GestureDetector.tsx#L543-L558) and [here](https://github.com/software-mansion/react-native-gesture-handler/blob/242c9af654428be7577067c8fd3c760e202008ac/ios/RNGestureHandlerModule.mm#L100-L101)). That's because Gesture Handler on iOS attaches `UIGestureRecognizer` to underlying `UIView`s and simply they must exist.

The main problem we have now is that when animating layout props, React and Reanimated overwrite each other's changes. That's because if Reanimated commits new ShadowNodes, the JS part of React still holds references to old ShadowNodes in Fiber nodes and clones them. Some time ago we've managed to get this working but it required modifying RN internals in a way that could affect the performance. We have a few ideas on how this could be achieved possibly without changes to RN core but need some time to test them out this week.

**Reply by ShikaSD** [reply]
> The main problem we have now is that when animating layout props, React and Reanimated overwrite each other's changes.

(Just thinking aloud here) Would it be possible to use State updates for those nodes instead to override properties? I am not sure if Reanimated uses built-in or custom views, but state was created specifically for merging React prop updates and native changes together

**Reply by ShikaSD** [reply]
> For example, in React Native Gesture Handler we also need to access ShadowNode in order to see if the corresponding view will get flattened by checking ShadowNodeTraits::FormsStackingContext

I know it doesn't solve the underlying problem, but maybe for this case it makes sense to enforce unflattening of nodes gesture recognizer is attached to (with `collapsable=false`? That's what we do for different event handlers internally as well.

**Reply by ShikaSD** [reply]
> JS VM that is used only from the UI thread to synchronously run user-defined JS functions (so-called worklets). 

What kind of information do you pass about the node there? How the worklet knows which props to interpolate/which node those updates belong to?

**Reply by tomekzaw** [reply]
> Would it be possible to use State updates for those nodes instead to override properties? I am not sure if Reanimated uses built-in or custom views, but state was created specifically for merging React prop updates and native changes together

Reanimated uses built-in RN `View` native component but wrapped with some additional JS logic ([here](https://github.com/software-mansion/react-native-reanimated/blob/688a244b3826b0f6cde52e9490eee5fe0d3e9bf9/src/createAnimatedComponent.tsx#L177-L699)) using `createAnimatedComponent` function ([here](https://github.com/software-mansion/react-native-reanimated/blob/688a244b3826b0f6cde52e9490eee5fe0d3e9bf9/src/reanimated2/component/WrappedComponents.ts#L5)) and exported from `react-native-reanimated` as `Animated.View`. 

We could implement `Animated.View` as a native component that extends the native part from RN's `View` using inheritance and just overwrite some of the methods, but we would still need to copy large chunks of JavaScript code from RN and we would like to avoid it if possible.

I need some more time to familiarize myself with the concept of `State` (I thought it only represents some internal state of native views like offset for `ScrollView`) but it sounds promising as it's kept in `ShadowNodeFamily`, can be updated using `setMostRecentState` and also is properly synchronized. Can you please show me where "merging React prop updates and native changes together" happens in the code?

> I know it doesn't solve the underlying problem, but maybe for this case it makes sense to enforce unflattening of nodes gesture recognizer is attached to (with `collapsable=false`? That's what we do for different event handlers internally as well.

We already set this property [here](https://github.com/software-mansion/react-native-gesture-handler/blob/a22374ea60035d0071ccffae91d86ceb8a30aa74/src/handlers/gestures/GestureDetector.tsx#L585) and it works well when the child of `<GestureDetector />` is a native component (e.g. `<View />`). But when the child is a functional component, I don't think we have a way to set `collapsable={false}` for the most outer component inside (if this is somehow possible, please let us know). That's why we only detect such cases using shadow tree (since from React ref we can obtain the ShadowNode of the most outer native component from the functional component) and tell users to set `collapsable={false}` on their own.

We cannot wrap the functional component within `<View collapsable={false} />` automatically because of two reasons. Firstly, if the functional component has a margin, then if we attach the gesture handler to the wrapper view, the gesture will also be able to capture the touches on the area of margin. Secondly, we think there may be some issues with position of the component when using `position: relative`.

> What kind of information do you pass about the node there?

On Paper, we identify the component by its viewTag, which we obtain from React ref in the main JS runtime. On Fabric, we do the same but using `ShadowNodeWrapper` instead of view tags.

> How the worklet knows which props to interpolate/which node those updates belong to?

[Here](https://docs.swmansion.com/react-native-reanimated/docs/fundamentals/animations#shared-value-animated-transitions)'s a short example from Reanimated docs. Firstly, you need to call `useAnimatedStyle` hook to define animated styles depending on shared values. Shared value can be written and read from both JS VMs. You can update its value immediately using assignment operator (e.g. `offset.value = 100;`), interpolate it (e.g. `offset.value = withSpring(100, { duration: 300 });` or calculate the value based on other shared values (e.g. `offset.value = 100 * foo.value + 200;`). The callback in `useAnimatedStyle` needs to be a worklet since this function will be called in the secondary JS context synchronously from the UI thread. Also, you can skip the `'worklet';` directive here as Reanimated's Babel plugin automatically recognizes `useAnimatedStyle` callbacks as worklets. Once you have "animated styles", you can simply pass them to `style` prop of `Animated.View` component, like here: `<Animated.View style={[styles.box, animatedStyles]} />`.

Reanimated first calculates the new values of all shared values (using so-called "mappers" which sort shared values topologically according to dependencies between them) and then detects which animated styles are affected. This part of Reanimated has already been implemented using JSI and works well both on Paper and Fabric (as it's arch-generic).

Now, for each affected animated style, [this](https://github.com/software-mansion/react-native-reanimated/blob/3106c8e0270c336ed39a0523ea10564e23e860b2/Common/cpp/NativeModules/NativeReanimatedModule.cpp#L377-L380) method gets called with `ShadowNode` and new styles as plain JS object. The thing we don't know yet is how to safely update the props and re-calculate layout so that RN won't overwrite these changes on next render.

**Reply by ShikaSD** [reply]
> Can you please show me where "merging React prop updates and native changes together" happens in the code?

I think it is a result of unfortunate wording, the props and state are kinda separate, and as you pointed out, we keep the track of the most recent state updates in the `ShadowNodeFamily`. What I meant by "merging" prop and native changes together is the fact that `ShadowNode` always has the most recent state even for stale nodes, meaning that it will see the state update no matter what. I don't think we change props directly through this mechanism though.
